### ISSUE TITLE
PopupMenuButton.onSelected is not called after the screen is rebuilt with an open menu

### ISSUE THREAD

**Issue by BurningAXE** [flutter/flutter#79873]
## Steps to Reproduce
Let me start by stating that I am not sure if this is really a bug or my understanding of Flutter's rebuild process is incomplete. So, with that in mind here are the steps:

1. Have an app with a PopupMenuButton with a few items
2. Open the popup menu
3. Have an enclosing ChangeNotifierProvider trigger a rebuild by calling notifyListeners()
4. Tap an item from the menu

**Expected results:** 
PopupMenuButton.onSelected is called

**Actual results:** 
PopupMenuButton.onSelected is not called

<details>
Doing some investigation I found the tap handling stops at this line:

[https://github.com/flutter/flutter/blob/4bf26b68014e0c2a2709d97b9c6b3222bdc2f848/packages/flutter/lib/src/material/popup_menu.dart#L1130](https://github.com/flutter/flutter/blob/4bf26b68014e0c2a2709d97b9c6b3222bdc2f848/packages/flutter/lib/src/material/popup_menu.dart#L1130)

Obviously the problem is that at the point of time I tap the menu item PopupMenuButtonState is not mounted. The question is - why isn't it? If the menu is shown on the screen shouldn't its _element be set to the corresponding element from the tree?

[✓] Flutter (Channel stable, 2.0.4, on macOS 11.2.3 20D91 darwin-x64, locale en-BG)
    • Flutter version 2.0.4 at /Applications/flutter
    • Framework revision b1395592de (5 days ago), 2021-04-01 14:25:01 -0700
    • Engine revision 2dce47073a
    • Dart version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/milen/Library/Android/sdk
    • Platform android-30, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

• No issues found!

</details>

**Comment by TahaTesser** [comment]
Hi @BurningAXE 
Thanks for filing the issue, here I created a simple code sample to test the issue, simply putting setState to rebuild, onSelected is called every time

Feel free to edit the code sample to make the issue reproducible and provided updated 

<details> 
<summary>code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

enum WhyFarther { harder, smarter, selfStarter, tradingCharter }

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  WhyFarther _selection;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Material App Bar'),
      ),
      body: Center(
        child: PopupMenuButton<WhyFarther>(
          onSelected: (WhyFarther result) {
            print(result);
            setState(() {
              _selection = result;
            });
          },
          itemBuilder: (BuildContext context) {
            setState(() {});
            return <PopupMenuEntry<WhyFarther>>[
              const PopupMenuItem<WhyFarther>(
                value: WhyFarther.harder,
                child: Text('Working a lot harder'),
              ),
              const PopupMenuItem<WhyFarther>(
                value: WhyFarther.smarter,
                child: Text('Being a lot smarter'),
              ),
              const PopupMenuItem<WhyFarther>(
                value: WhyFarther.selfStarter,
                child: Text('Being a self-starter'),
              ),
              const PopupMenuItem<WhyFarther>(
                value: WhyFarther.tradingCharter,
                child: Text('Placed in charge of trading charter'),
              ),
            ];
          },
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.add),
        onPressed: () {},
      ),
    );
  }
}

``` 

</details> 


<details> 
<summary>flutter doctor -v</summary> 

```bash[✓] Flutter (Channel stable, 2.0.4, on macOS 11.2.3 20D91 darwin-x64, locale en-GB)
    • Flutter version 2.0.4 at /Users/tahatesser/Code/flutter_stable
    • Framework revision b1395592de (6 days ago), 2021-04-01 14:25:01 -0700
    • Engine revision 2dce47073a
    • Dart version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.55.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (3 available)
    • iPhone 12 (mobile) • 83060656-28E7-44CD-801E-B11EC3EC89BA • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-4
      (simulator)
    • macOS (desktop)    • macos                                • darwin-x64     • macOS 11.2.3 20D91 darwin-x64
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 89.0.4389.114

• No issues found!

```
	
</details>

**Comment by BurningAXE** [comment]
@TahaTesser Thanks for the sample! I tried to reproduce it and couldn't! After much debugging and adding elements from my app to it I found that it was indeed my code that caused the issue... I was incorrectly recreating a `GlobalKey` of a parent widget on each rebuild. I guess it is expected that this will severed the `PopupMenuButton`'s head!? If so, the issue is good to be closed and I apologise for the wasted time!
